cloudevents/spec.md

```diff
@@ -153,7 +153,8 @@ together into a transport envelope body.
 
 Every CloudEvent conforming to this specification MUST include context
 attributes designated as REQUIRED, MAY include one or more OPTIONAL context
-attributes and MAY include one or more extension attributes.
+attributes and MAY include one or more extension attributes. Each context
+attribute MUST only appear at most once in a CloudEvent.
```

Comment:
Shouldn't we require that for extension attributes as well?

Reply:
If not, I think we may have different mental models that would be worth discussing. My world view is:

A CloudEvent has attribute. Each attribute has a name, a type, and may have some validation constraints. Each attribute is either:

- Required by the declared spec version
- Optional but specified by the declared spec version
- An extension attribute, i.e. unspecified in the declared spec version

Typically "context attribute" is used to mean the first two but does not include extension attributes. Does that match your view? Maybe we should call this out and have a single term ("event attribute"? just "attribute") which can be applied whenever we mean all three. Or maybe "context attribute" should be specified to include them all?

Reply:
I think we have the same mental model, and the spec does say:
- - -
A CloudEvent MAY include any number of additional context attributes with distinct names, known as "extension attributes". Extension attributes MUST follow the same [naming convention](https://github.com/cloudevents/spec/blob/44c9579dbe09d1b8f25a7655fc480e7f6d6b8796/cloudevents/spec.md#naming-conventions) and use the same [type system](https://github.com/cloudevents/spec/blob/44c9579dbe09d1b8f25a7655fc480e7f6d6b8796/cloudevents/spec.md#type-system) as standard attributes. 
- - -
which to me means extensions attributes are context attributes. But since it wasn't clear to you in this new sentence, there it might not be clear to others too. I modified the new sentence and tweaked the previous sentence to it even more clear that extensions attributes ARE context attributes too. See if that helps.

Reply:
Actually, re-reading your comment:
> Typically "context attribute" is used to mean the first two but does not include extension attributes.

I disagree with this part :-)  I think context attributes include all 3 - as the spec implies in the text I quoted.

Reply:
Great. This all sounds good - I'll see if I can find where in the spec I got that impression from, or whether it was just a misreading.

Reply:
At a brief look, I think this sentence in spec.md could be changed very slightly:

> Every CloudEvent conforming to this specification MUST include context attributes designated as REQUIRED, MAY include one or more OPTIONAL context attributes and MAY include one or more extension attributes.

Just changing "extension attributes" to "extension context attributes" would help a lot.

The primer could do with a bit more work in terms of consistency - it talks about "CloudEvent Attribute Extensions" but only uses "context attribute" as a term in the versioning part. Obviously the primer isn't intended to be as formal as the main spec, but some consistency would be good.

I did find a term that I think would be good to use more widely when we mean "the attributes already in the spec" - "standard attributes".

Let me know if you want me to have a pass at clarifying things.

Reply:
> Just changing "extension attributes" to "extension context attributes" would help a lot.

Yup - that was part of my last commit. I just push a tweak to the Primer to make the wording a bit more consistent